### PR TITLE
feat(Java): Guess types when we can't resolve them definitively

### PR DESCRIPTION
**Pull request by nmote**
We currently use type inference to handle typed metavariables. For example, a Java pattern might include `(boolean $X)` which will match any expression that is inferred to have the type `boolean`.

This PR extends type inference to include heuristics which we use to make a best guess for types of expressions whose types we cannot infer using traditional means. This typically happens when there is an unresolved name involved in the type.

While the heuristics are based on naming patterns in the Java standard library, they are not limited to the standard library. Nor do we intend to replicate the entire Java standard library as heuristics. Instead, we should include heuristics when we believe that they correctly describe the type of an expression with high confidence, whether they are based on a convention in the standard library or other common conventions.

Test plan: Automated tests

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: tests/patterns/java/metavar_typed_bool.java [hunk 1]
```diff
@@ -68,4 +71,40 @@ public static void overloaded() {
...
+
+    Map<String, String> m1 = new HashMap<>();
+    m1.put("x", "y");
+    // OK:
+    System.out.println(m1.get("x"));
+
+    Map<String, Boolean> m2 = new HashMap<>();
+    m1.put("x", false);
+    // MATCH:
+    System.out.println(m2.get("x"));
```
Comment (IagoAbal) [inline tests/patterns/java/metavar_typed_bool.java:108]: Does thing handle Java for-each loops or will that go into another PR?

**Review by IagoAbal — APPROVED** [review]

File: tests/patterns/java/metavar_typed_bool.java (hunk 1 above)
Comment (IagoAbal) [reply tests/patterns/java/metavar_typed_bool.java:108]: Also need to handle `getX` methods. Although maybe this probably needs to be done in Pro.

Comment (nmote) [reply tests/patterns/java/metavar_typed_bool.java:108]: No, it doesn't handle for-each loops. And yeah, I think the `getX` heuristics need to go into Pro since we need detailed information about the resolved type, not just its name.
